src/plots/barPlot.ts

```diff
@@ -948,9 +945,182 @@ function computeBarPixelThickness(
+    const myKeys = Object.keys(this.obj);
+    const otherKeys = Object.keys(other.obj);
+
+    if (myKeys.length !== otherKeys.length) {
+      return true;
+    }
+
+    for (const key of myKeys) {
+      if (!other.obj.hasOwnProperty(key)) {
+        return true;
+      }
+      if (this.obj[key].isDifferent(other.obj[key])) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+function signRef(a: any) {
+  return new ReferenceSignature(a);
+}
+
+function signArray(a: any[]) {
+  return new ArraySignature(a.map((element) => sign(element)));
+}
+
+function signObj(obj: { [key: string]: any }) {
+  const signatureRecord: ISignatureRecord = {};
+  for (const key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      signatureRecord[key] = sign(obj[key]);
+    }
+  }
+  return new ObjectSignature(signatureRecord);
+}
+
+function signScale(scale: Scale<any, any>) {
+  const scaleObj = {
+    domain: scale.domain(),
+    range: scale.range(),
+    updateId: scale.updateId(),
+    ref: signRef(scale),
+  };
+  return signObj(scaleObj);
+}
+
+function signDataset(dataset: Dataset) {
+  const datasetObj = {
+    ref: signRef(dataset),
+    // only sign updateId since only data() and metadata() exist as properties
+    // and both update updateId
+    updateId: dataset.updateId(),
+  };
+  return signObj(datasetObj);
+}
+
+function sign<T>(a: T): Signature {
+  if (a instanceof Signature) {
+    return a;
+  } else if (a instanceof Scale) {
+    return signScale(a);
+  } else if (a instanceof Dataset) {
+    return signDataset(a);
+  } else if (a !== null && typeof a === "object") {
+    return signObj(a);
+  } else if(Array.isArray(a)) {
+    return signArray(a);
+  } else {
+    return signRef(a);
+  }
+}
+
+type LockedMemoize<F extends Function> = F & {
+  /**
+   * Lock the memoization to always return the cached
+   * property during the duration of fn. This lets you
+   * bypass the performance hit of signing when you
+   * know the fn will not mutate the inputs.
+   *
+   * Be sure to force the memoization to the value
+   * you want before calling this!
+   *
+   * @param fn
+   */
+  doLocked<T>(fn: () => T): T;
+};
+
+function memoize<F extends Function>(compute: F): LockedMemoize<F> {
+  let lastSignature: Signature = undefined;
+  let lastValue: any;
+  let locked = false;
+
+  const memoizeFn = function(...args: any[]) {
+    if (locked) {
+      return lastValue;
+    }
+    const inputSignature = signArray(args);
+    if (lastSignature === undefined
+        || lastSignature.isDifferent(inputSignature)) {
+      lastSignature = inputSignature;
+      lastValue = compute.apply(this, args);
+    }
+    return lastValue;
+  } as any as LockedMemoize<F>;
+  memoizeFn.doLocked = function<T>(cb: () => T) {
+    if (locked) {
+      throw new Error("Locking an already locked memoize function!");
+    }
+    locked = true;
+    const retVal = cb.apply(this);
+    locked = false;
+    return retVal;
+  };
+  return memoizeFn;
```

Comment:
this is a lot of new code that is fairly complex. would be amazing to get any sort of test for it.